src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -680,23 +686,152 @@ public new CSharpCompilation AddSyntaxTrees(IEnumerable<SyntaxTree> trees)
+                this.Options,
+                seenFiles,
+                this.MessageProvider,
+                IsSubmission,
+                ordinalMapBuilder,
+                ref loadDirectiveMap,
+                ref declMap,
+                ref declTable,
+                ref referenceDirectivesChanged);
+            seenFiles.Free();
+        }
+
+        /// <summary>
+        /// Recursively adds trees #load'ed by <paramref name="tree"/> to the <paramref name="treesBuilder"/>.
+        /// </summary>
+        private static void AppendSyntaxTreesHelper(
+            ArrayBuilder<SyntaxTree> treesBuilder,
+            SyntaxTree tree,
+            ParseOptions parseOptions,
+            CSharpCompilationOptions options,
+            HashSet<string> seenFiles,
+            CommonMessageProvider messageProvider,
+            bool isSubmission,
+            IDictionary<SyntaxTree, int> ordinalMapBuilder,
+            ref ImmutableDictionary<SyntaxTree, ImmutableArray<LoadDirective>> loadDirectiveMap,
+            ref ImmutableDictionary<SyntaxTree, Lazy<RootSingleNamespaceDeclaration>> declMap,
+            ref DeclarationTable declTable,
+            ref bool referenceDirectivesChanged)
+        {
+            // It's okay if tree.FilePath is null or empty, because we only expect to see a single tree
+            // with no path information (the initial "code" passed to the CSharpScript constructor).
+            seenFiles.Add(tree.FilePath);
+
+            var resolver = options.SourceReferenceResolver;
+            ArrayBuilder<LoadDirective> loadDirectives = null;
+            foreach (var directive in tree.GetCompilationUnitRoot().GetLoadDirectives())
+            {
+                SyntaxTree loadedTree = null;
+                var diagnostics = DiagnosticBag.GetInstance();
+
+                var fileToken = directive.File;
+                var path = (string)fileToken.Value;
+                if (path == null)
+                {
+                    // If there is no path, the parser should have some Diagnostics to report.
+                    Debug.Assert(tree.GetDiagnostics().Any());
+                    continue;
+                }
+
+                var resolvedFilePath = resolver.ResolveReference(path, baseFilePath: null);
+                if (resolvedFilePath == null)
+                {
+                    diagnostics.Add(
+                        messageProvider.CreateDiagnostic(
+                            (int)ErrorCode.ERR_NoSourceFile,
+                            fileToken.GetLocation(),
+                            path,
+                            CSharpResources.CouldNotFindFile));
+                }
+                else if (!seenFiles.Contains(resolvedFilePath))
                 {
-                    throw new ArgumentException(CSharpResources.SubmissionCanHaveAtMostOne, "trees");
+                    var readTextDiagnostics = new List<Diagnostic>();
```

Comment:
Per suggestion from Tomas, I will make ReadText throw an Exception and this List will go away.